subprojects/core/src/main/java/org/gradle/execution/DefaultWorkValidationWarningRecorder.java

```diff
@@ -34,11 +35,11 @@
 public class DefaultWorkValidationWarningRecorder implements ValidateStep.ValidationWarningRecorder, WorkValidationWarningReporter {
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultWorkValidationWarningRecorder.class);
 
-    private final AtomicInteger workWithFailuresCount = new AtomicInteger();
+    private final Set<UnitOfWork> workWithWarnings = ConcurrentHashMap.newKeySet();
 
     @Override
     public void recordValidationWarnings(UnitOfWork work, Collection<? extends InternalProblem> warnings) {
-        workWithFailuresCount.incrementAndGet();
+        workWithWarnings.add(work);
```

Comment:
🤔 It look suspicious that we are now holding references to the units of work until the end of the build. Should we hold their identity instead?

Reply:
That's a good point, I made the change.